### ISSUE TITLE
should_implement_trait should not warn async methods

### ISSUE THREAD

**Issue by taiki-e** [rust-lang/rust-clippy#4290]
They are actually returning a different type than normal methods.

```rust
#![feature(async_await)]
#![warn(clippy::should_implement_trait)]

pub struct Foo;

impl Foo {
    // actual type: pub fn next(&mut self) -> impl Future<Output = Option<()>>
    pub async fn next(&mut self) -> Option<()> { //~ WARN: defining a method called `next` on this type; consider implementing the `std::iter::Iterator` trait or choosing a less ambiguous name
        unimplemented!()
    }
}
```

[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=bd3ba14595a4d8821e23ea013c451a10)

**Comment by flip1995** [comment]
The warning is mostly about the method name `next`. Since this is a style lint, I think it is totally fine, that the lint triggers here. Maybe calling the function `next_async` would be better here?

**Comment by djc** [comment]
I kind of disagree, since the `Stream` trait actually also calls its method `next()`. If `Stream` was in `std`, this lint could refer to `Stream`, but as long as that's not the case, clippy would be better off not warning about this IMO (I've been hit by this multiple times and always just end up silencing the lint).

**Comment by phansch** [comment]
This was apparently fixed by https://github.com/rust-lang/rust-clippy/pull/5437
